### ISSUE TITLE
[Enhancement] Add options for own preparsers

### ISSUE THREAD

**Issue by GitMensch** [OpenCobolIDE/OpenCobolIDE#137]
OCIDE calls cobc for compilation and conditional dbpre/esqlOC for specific files.
I'd like to see a more general approach being able to set up different preparsers (or even compilers) for specific file types.

Theses are the parts to consider:

* [optional, but nice] parser (to "know" about special sections/paragraphs that don't look like sections/paragraphs, and in the case I'd like to configure these have ";" in column 7 which may lead to not parse the files correctly) - can this part be "inserted" at runtime without the need to recompile the application at all?
* preferences - should be easily doable if the current SQL COBOL preference is changed to "external compilers" and has a list to choose the compilers, "hiding" the current text and file selection list until an entry is added/edited. the default would be only one entry for SQL COBOL (see below)

    Name: SQL COBOL
    Associated extensions: .sqb
    Compiler path: [empty, has to be filled with esql binary)
    Compiler flags: 
    [Checkbox] Only Preparser           Output file [only active if preparser is checked]:  [whatever you do currently]

If "Only Preparser" is checked the output file would be passed to cobc, if not the compiler either does the compilation on it's own or it calls cobc itself.

**Comment by ColinDuquesnoy** [comment]
This is too much for where I want to lead OpenCobolIDE but this is definitely something I will think about for the next IDE I talked you about. 

For me the biggest issue at the moment is that the compiler settings are applied globally on any file, there is no way to build a simple hello world module without linking to any 3rd party libraries and build another file that link to e.g. gtk without changing the compiler setup. This may be very annoying... In the next IDE, I plan to give the user the ability to be able to configure each file independently and each compiler will have an editable default configuration that is used to pre-fill the file configuration.

> [optional, but nice] parser (to "know" about special sections/paragraphs that don't look like sections/paragraphs, and in the case I'd like to configure these have ";" in column 7 which may lead to not parse the files correctly) - can this part be "inserted" at runtime without the need to recompile the application at all?

So If I understand well, you'd like to be able to mark some lines in the document (by putting a ';' in column 7) and have them appear in the navigation panel, right? If yes, that should be fairly easy to do and yes this will show up at runtime as soon as you've stopped typing. *If we have a starting and closing characters (e.g. ; and :), we could even make them appear in the fold tree and collapse anything that is between those two characters.*

> should be easily doable if the current SQL COBOL preference is changed to "external compilers" and has a list to choose the compilers, "hiding" the current text and file selection list until an entry is added/edited. the default would be only one entry for SQL COBOL (see below)

Things are a bit more complicated... E.g. for dbpre to work, I need to  copy some files next to the sources and call cobc a bit differently (need to pass cobmysqlapi.o and link with mysql)... For esqLOC to work I need setup some environment variables, run vcvarsall.bat,...

* esqloc:

```
    esqlOC.exe -static -o file.cob file.sqb

    GnuCobol Commands:
    SET OC_RUNTIME=c:\OpenCobol_bin
    SET esqlOC_RUNTIME=c:\esqlOC\release
    SET COB_CFLAGS=-I %OC_RUNTIME%
    SET COB_LIBS=
      %OC_RUNTIME%\libcob.lib %OC_RUNTIME%\mpir.lib %esqlOC_RUNTIME%\ocsql.lib
    SET COB_CONFIG_DIR=
      %OC_RUNTIME%\config\ set PATH=C:\WINDOWS\system32;%OC_RUNTIME%
    call "%PROGRAMFILES%\Microsoft Visual Studio 10.0\VC\vcvarsall.bat"
    %OC_RUNTIME%\cobc.exe -fixed -v -x -static -o c:\Temp\esqlOCGetStart1.exe
      c:\Temp\esqlOCGetStart1.cob
```

* dbpre:

```
/path/to/dbpre SOURCE.scb -I/path/to/framework -ts=TAB_LEN
copy PGCTBBATWS to source dir
cobc -x SOURCE.cob /path/to/cobmysqlapi.o -L/usr/lib/mysql
-lmysqlclient -o bin/SOURCE.exe
```

All those things are not easy to setup through an interface... What do you think?

> If "Only Preparser" is checked the output file would be passed to cobc, if not the compiler either does the compilation on it's own or it calls cobc itself.

I don't get this part... I probably misunderstood but for me, if "Only Preparser" is checked, I would not pass the output to cobc.

**Comment by GitMensch** [comment]
For the "Only Preparser" option:
It would be nice to either have OCIDE call the "compiler" and call cobc with the generated file afterwards (as it is done now) or to have the external compiler do all the compilation (call cobc itself). So you seem to have it understood what I've thought :-)

For the external compiler setup: I guess the esql stuff above is only a sample as all these flags should be changed to the compiler flags already in preferences dialog, you only need to add -l "%esqlOC_RUNTIME%\ocsql.lib". For dbpre it's an additional copy (what is this for, what does it copy?)

Leaving the last out and setting the rule "output has to be $file.cob" it's "only":
"options passed to Compiler" and  "additional options passed to COBOL compiler"

Name: SQL COBOL (esql)
Associated extensions: .sqb
Compiler path: [empty, has to be filled with path to esql binary, as it has to be done for own cobc configuration)
Compiler flags: -static -o $file.cob $file.sqb
Additional COBOL Compiler flags: -l "%esqlOC_RUNTIME%\ocsql.lib"
Only Preparser [checkbox enabled]

Name: SQL COBOL (dbpre)
Associated extensions: .scb
Compiler path: [empty, has to be filled with path to dbpre binary, maybe add the rule that cobmysalapi.o has to be placed, there, too, this would remove the need for changing the path in additional compiler flags)
Compiler flags: $file.scb -I/path/to/framework -ts=TAB_LEN
Additional COBOL Compiler flags: /path/to/cobmysqlapi.o -L/usr/lib/mysql -lmysqlclient
Only Preparser [checkbox enabled]

**Comment by ColinDuquesnoy** [comment]
Thanks for those clarifications, I can see clearly what you want now. I'll keep it mind for the new design ;)

**Comment by GitMensch** [comment]
Can we have a temporary workaround? Specify file type + batch/shell to call. This way the user has the option to run a n external compiler and cobc. Even if he has to setup everything in both OCIDE and the shell it's much better than the current option "is not available".

A quite good workaround would be to have two input fields [sample in brackets] and a checkbox (all default to empty)

file extension: [.myext]  compiler batch: [/path/to/mycomp.sh]  call for checks [X]
Compiler batch is called with some environment vars you define (something like OCIDE_FILE, OCIDE_INCLUDES, OCIDE_COBC_CALL, OCIDE_CHECKONLY]), only if clicked on "compile"/"run".
The backround process would only call it if "call for checks" is active (and set OCIDE_CHECKONLY=J" in this case)

It *looks* easy to implement and even if it's no "very clean and multi purpose solution" it's at least a "if you know what to do you can do anything solution".

What do you think about this?

**Comment by ColinDuquesnoy** [comment]
Yes should be fairly easy to implement. I'll see if I can make it before this weekend, otherwise this will get into the next development version (4.7). 

But I am not sure what is the meaning of OCIDE_INCLUDES and COBC_CALL. What should I set for those variables?

**Comment by GitMensch** [comment]
set OCIDE_COBC_CALL to the complete cobc line you'd call if it wasn't an "unknown preparser". The called batch can decide to use this or not.
OCIDE_INCLUDES could be a list of includes as separated list with the systems path separator, for my own setting this would result in OCIDE_INCLUDES=/path/for/includes:../cpy
The batch can use this to set COB_COPY_DIR (and extend it, or set other vars, or separate them or whatever).

**Comment by ColinDuquesnoy** [comment]
> set OCIDE_COBC_CALL to the complete cobc line you'd call if it wasn't an "unknown preparser". The called batch can decide to use this or not.

got it

> OCIDE_INCLUDES could be a list of includes as separated list with the systems path separator, for my own setting this would result in OCIDE_INCLUDES=/path/for/includes:../cpy
The batch can use this to set COB_COPY_DIR (and extend it, or set other vars, or separate them or whatever).

I see but where would those paths come from? Since the extra compiler settings can contains anything I cannot use this as a source. Maybe add a third field where you can input those paths?

**Comment by GitMensch** [comment]
No, just provide what is set in the preferences for the COBOL compiler to be able to use or skip this from the batch.

**Comment by ColinDuquesnoy** [comment]
I implemented this in hackedit. For the moment, I tested it with dbpre:

* Preparser configuration:
![dbpre-config](https://cloud.githubusercontent.com/assets/1681217/10417343/a028e6e8-7038-11e5-8e62-d02cf4252519.png)

* Compilation output:
![dbpre-compile](https://cloud.githubusercontent.com/assets/1681217/10417344/a38740a0-7038-11e5-84ac-ff0b32b2db9c.png)


I think I still need one or two week to test this feature on Windows with esql, add support for msvc based compiler and implement the last feature requested for OCIDE v4.7 in HackEdit then it  should be ready for a first public alpha release.

It's unlikely that I will backport this feature to OCIDE.
